include/deal.II/grid/manifold_tools.h

```diff
@@ -0,0 +1,105 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__manifold_tools_h
+#define dealii__manifold_tools_h
+
+
+/*----------------------------   manifold_tools.h     -----------------*/
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/base/thread_management.h>
+#include <deal.II/base/point.h>
+#include <deal.II/base/derivative_form.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/manifold.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * CompositionManifold.  Take two ChartManifold objects, and make
+ * their composition. The CompositionManifold object is a
+ * ChartManifold going from the chart of the first ChartManifold to
+ * the embedding space of the second ChartManifold.
+ *
+ * This class only works for dim <= chartdim <= spacedim1 <=
+ * spacedim. If you try to instantiate anything different, an
+ * Exception will be thrown in one of the ChartManifold classes that
+ * violates this condition.
+ *
+ * Given the ChartManifold F and the ChartManifold G, this class
+ * represents the composition of G after F.
+ *
+ * @ingroup manifold
+ * @author Luca Heltai, Timo Heister, 2016
+ */
+template <int dim, int spacedim=dim, int chartdim=dim,
+          int dim1=dim, int dim2=dim, int spacedim1=dim>
```

Comment:
dim1 and dim2 are less useful compared to spacedim1 (and more likely to be equal to dim), so I would move them to the end. Can you also rename spacedim1 to something better? Maybe "intermediate_dim" or something? It is probably worth documenting all of them with ``@tparam``.